From fb051f3bdb8aaab4f42f5c19fade791dbbabbdb7 Mon Sep 17 00:00:00 2001
From: magnum <john.magnum@hushmail.com>
Date: Fri, 2 Dec 2016 09:47:55 +0100
Subject: [PATCH] Fix a hideous bug in CPU version of WPAPSK format concerning
 new_keys flag. It was only set in clear_keys() but not in set_key(). This is
 not enough when running Single mode.

Also "fix" Single mode (cracker.c) so clear_keys() is always called when
"expected" - I am pretty sure some more jumbo formats depend on this.

Closes #2344. Any one of the fixes would suffice but I opted to do both
for now.
---
 src/cracker.c | 1 +
 src/wpapsk.h  | 1 +
 2 files changed, 2 insertions(+)

diff --git a/src/cracker.c b/src/cracker.c
index 33294e7..c3bbf16 100644
--- a/src/cracker.c
+++ b/src/cracker.c
@@ -1093,6 +1093,7 @@ int crk_process_salt(struct db_salt *salt)
 			if (!salt->list)
 				return 0;
 			index = 0;
+			crk_methods.clear_keys();
 		}
 	}
 
diff --git a/src/wpapsk.h b/src/wpapsk.h
index 59557c3..7a74787 100644
--- a/src/wpapsk.h
+++ b/src/wpapsk.h
@@ -323,6 +323,7 @@ static void set_key(char *key, int index)
 		length = PLAINTEXT_LENGTH;
 	inbuffer[index].length = length;
 	memcpy(inbuffer[index].v, key, length);
+	new_keys = 1;
 }
 
 static char *get_key(int index)
